Generate invoice

tochman opened the pull request:
### PT Story: https://www.pivotaltracker.com/story/show/150783276

* Adds functionality to generate and attach an invoice for an order
* Adds link to view invoice in new window
* Uploads invoice to S3
* Adds due date for payment once invoice has been generated
* Displays all the necessary info on order
* And then some...

<img width="605" alt="skarmavbild 2017-09-15 kl 00 02 07" src="https://user-images.githubusercontent.com/5248170/30457816-36152f9a-99a9-11e7-8a27-c58b3abd1f24.png">

## What I've learned
* That Prawn is badly documented but quite powerful
* That designing invoices is hard work
* That I probably should think twice before taking on 3 point features on student projects
* That staying in scope is easier said than done 😳
* That coding is really, really fun! 😎

tochman committed:
adds feature file with backgroud steps to set state

tochman committed:
removes debugger

tochman committed:
changes OrderItem settings to use SEK as default
skips _cents posfix on price table

tochman committed:
refactors step definitions
adds assertion for order taxes and total
modifies order view
updates scenario

j0sv commented on app/models/order.rb:
```diff
@@ -1,3 +1,8 @@
 class Order < ApplicationRecord
   acts_as_shopping_cart_using :order_item
+  SWEDISH_VAT = 12.5
```
The vat is 12%.

tochman committed:
currency settings done
updates index view with humanized totals and subtotals
refactors scenarios

tochman committed:
adds missing migration used to reset currency values

tochman committed:
adds gems for atachments and pdf generation

tochman committed:
adds attachment class with association to order
adds paperclip gem
adds and configures factory

tochman committed:
adds blank pdf fixture

tochman committed:
adds content validation to attachment
fixes factory for attachment

tochman committed:
adds route for generating invoices
adds basic invoice content

tochman committed:
WIP changes migrations and adds ttf font to generated invoice

tochman committed:
adds s3 storage to attachment - remember to add aws mock in test env

tochman committed:
adds steps to analyze pdf file

tochman committed:
going green on pdf content

tochman committed:
adds tmp folder to version control to stop ci from complaining

tochman committed:
changes to Rails.root in step def and starts to loose my patiance

tochman committed:
moves pdf generator to separate service

satorinic commented on features/admin_manages_invoices.feature:
```diff
@@ -0,0 +1,46 @@
+      | Dish 2 | Description for Dish 2 | 200   | 10           | Starter  |
+      | Dish 3 | Description for Dish 3 | 300   | 10           | Starter  |
+
+    And the following orders exist:
```
Don't know if it matters, but in the database we have two separate columns for time and date.

satorinic commented on features/step_definitions/order_steps.rb:
```diff
@@ -14,3 +14,29 @@
+
+
+Given(/^"([^"]*)"'s order contains:$/) do |billing_name, table|
+  @order = Order.find_by(billing_name: billing_name)
```
isn't better to create an `billing_id`and use that instead of using `billing_name`?

tochman replied:
In test? I don't see a need for that

satorinic commented on features/step_definitions/order_steps.rb:
```diff
@@ -14,3 +14,29 @@
+end
+
+Then /^the subtotal for the cart should be "([^"]*)"$/ do |subtotal|
+  @order = @order || Order.last
```
Isn't better to create an `order.id`(corresponding to my comment on line 20?)or something similar?

tochman replied:
This is just for testing purposes.

satorinic commented on app/services/pdf_generator_service.rb:
```diff
@@ -0,0 +1,42 @@
+      pdf.font Rails.root.join('app', 'assets', 'fonts', 'futura.ttf')
+      pdf.text @order.billing_name, size: 14
+      pdf.move_down 16
+      pdf.text "Tax (VAT): #{helpers.humanized_money_with_symbol @order.taxes}", size: 14
```
great to include "moms"!

tochman committed:
WIP invoice design

tochman committed:
WIP invoice design - 2

tochman committed:
still messing around with the tables... documentation suck big time...

tochman committed:
Still WIP

tochman committed:
adds due_date to Order
adds instance method to set due date
keep the logic out of the controlers. Yay

tochman committed:
adds fonts and finalizes v1 of invoice design
conditions display of 'View Invoice' link
displays generated invoice in new window
adds instance method #has_invoice? to Order

tochman committed:
Merge branch 'develop' into generate_invoice

tochman committed:
missed to remove unused scope...

tochman committed:
adds exception if order has no items on invoice generation

tochman committed:
Merge branch 'develop' into generate_invoice
